src/Morphir/Scala/Feature/Codec.elm

```diff
@@ -478,50 +470,254 @@ genEncodeReference tpe =
+            Scala.Apply (Scala.Variable scalaFqn) yeildArgValues
+    in
+    if List.isEmpty ctorArgs then
+        ( Scala.LiteralMatch (Scala.StringLit (ctorName |> Name.toTitleCase))
+        , Scala.Apply (Scala.Variable "Right")
+            [ Scala.ArgValue Nothing <| Scala.Variable scalaFqn
+            ]
+        )
+            |> Ok
+
+    else
+        generatorsResult
+            |> Result.map
+                (\generators ->
+                    ( Scala.LiteralMatch (Scala.StringLit (ctorName |> Name.toTitleCase))
+                    , Scala.ForComp generators yeildExpression
+                    )
+                )
 
 
 {-|
 
     Get an Decoder reference for a Type
 
 -}
-genDecodeReference : FQName -> Type ta -> Result Error Scala.Value
-genDecodeReference fqName tpe =
+mapTypeToDecoderReference : Maybe ( Name, Scala.Path ) -> Type ta -> Result Error Scala.Value
+mapTypeToDecoderReference maybeTypeNameAndPath tpe =
     case tpe of
         Type.Variable _ varName ->
-            Ok (Scala.Variable ("decode" :: varName |> Name.toCamelCase))
+            Scala.Variable ("decode" :: varName |> Name.toCamelCase)
+                |> Ok
 
         Type.Reference _ ( packageName, moduleName, typeName ) typeArgs ->
             let
                 scalaPackageName =
-                    packageName ++ moduleName |> List.map (Name.toCamelCase >> String.toLower)
+                    packageName
+                        ++ moduleName
+                        |> List.map (Name.toCamelCase >> String.toLower)
 
-                scalaModuleName =
-                    "Codec"
+                codecPath : List String
+                codecPath =
+                    List.concat [ scalaPackageName, [ "Codec" ] ]
 
-                scalaReference =
-                    Scala.Ref
-                        (scalaPackageName ++ [ scalaModuleName ])
-                        ("decode" :: typeName |> Name.toCamelCase)
+                decoderName : String
+                decoderName =
+                    "decode" :: typeName |> Name.toCamelCase
+
+                scalaReference : List String -> String -> Scala.Value
+                scalaReference path name =
+                    Scala.Ref path name
             in
-            Ok scalaReference
+            case typeArgs of
+                [] ->
+                    Ok (scalaReference codecPath decoderName)
+
+                _ ->
+                    typeArgs
+                        |> List.map (\typeArg -> mapTypeToDecoderReference maybeTypeNameAndPath typeArg)
+                        |> ResultList.keepFirstError
+                        |> Result.map (List.map (Scala.ArgValue Nothing))
+                        |> Result.map (Scala.Apply (scalaReference codecPath decoderName))
 
         Type.Tuple a types ->
             let
+                tupleDecoderRef : Scala.Value
+                tupleDecoderRef =
+                    Scala.Ref [ "morphir", "sdk", "tuple", "Codec" ] "decodeTuple"
```

Comment:
Same thing here. Shouldn't use the SDK decoder for tuples but a specific inline implementation like in the case of records.